Add support for Git Credential Manager (beta-only)

niik opened the pull request:
<!--
What GitHub Desktop issue does this PR address? (for example, #1234)
-->

Closes https://github.com/desktop/desktop/issues/18654
Ref https://github.com/desktop/dugite-native/pull/510
Ref https://github.com/desktop/dugite-native/pull/508

## Description
<!--
A summary of the changes made along with any other information that would be helpful to a reviewer such as potential tradeoffs or alternative approaches you considered.
-->

This adds experimental (beta only) support for [Git Credential Manager](https://github.com/git-ecosystem/git-credential-manager) (GCM) which we're now bundling inside of GitHub Desktop's custom Git environment.

When enabling GCM GitHub Desktop will forward all requests for credentials that don't match accounts stored in GitHub Desktop (in essence third party hosts such as Azure Devops, BitBucket, etc).

GCM currently supports [Azure DevOps](https://azure.microsoft.com/en-us/products/devops), Azure DevOps Server (formerly Team Foundation Server), Bitbucket, GitHub, GitLab, and generic username + password hosts.

The feature is currently disabled by default and limited to ~~beta~~development.

Note for reviewers: Hiding whitespace can should help with reviewing changes in git/core.ts.

### Screenshots

<!--
If this PR touches the UI layer of the app, please include screenshots or animated gifs to show the changes.
-->

<img width="604" src="https://github.com/desktop/desktop/assets/634063/34517d3e-ee91-4af8-aaf2-e6917f364c4e" alt="Screenshot of Desktop's advanced preferences section with a highlight around the new option to enable Git Credential Manager" />

## Release notes

<!--
You can leave this blank if you're not sure.
If you don't believe this PR needs to be mentioned in the release notes, write "Notes: no-notes".
-->

Notes: [New] Option to use Git Credential Manager for repositories hosted outside of GitHub.com

niik committed:
Support GCM credential helper for third-party credentials

niik committed:
:art: cleanup

niik committed:
Needs a lil slash

niik committed:
Whoopsie

niik committed:
Let's just hardcode manager for now

niik committed:
Pass the path down to the trampoline environment

We need to run the credential helper in the repo path in case there's local config variables at play

niik committed:
Put GCM usage behind a preference for now

niik commented:
Seems we've got some legit test failures here. One of them is related to our use of '--keep-reduntant-commits' in

https://github.com/desktop/desktop/blob/afb7d2edca738be44635b007bde44a8ba68b4720/app/src/lib/git/cherry-pick.ts#L160-L173

`Error: GitError: fatal: cherry-pick: --keep-redundant-commits cannot be used with --continue`

This was added in git v2.45.1: https://github.com/git/git/commit/bd2f9fd025e3aa2926f249e9c24f1f648e89d997 and I have not dug further into it yet. @tidy-dev do you recall the details around this parameter?

The second issue seems to be that our localStorage mock isn't working as intended. Looking into that

tidy-dev commented:
> --keep-rundandant-commits

Here is some context: https://github.com/desktop/desktop/pull/11717

I haven't tested but I believe if you say had commits: A B C D E on a branch and attempted to cherry-pick B C D to another branch where C would be redundant and thus result in an empty commit, it would stop progression and through and error. Apparently the other option was to skip the commits, we went with keep redundant so that commits didn't just disappear. I.E., User cherry-picks 3 but only 2 show up in target branch. A user might be confused about where it went.  Possible third option that I didn't explore.. would be capture that error, and putting in the success banner "Redundant commits were skipped"

tidy-dev commented:
Looks like `--keep-redundant-commits` was deprecated. 

https://git-scm.com/docs/git-cherry-pick#Documentation/git-cherry-pick.txt---keep-redundant-commits

Maybe we can just swap for `--empty=keep` - [It would appear so from the documentation](https://git-scm.com/docs/git-cherry-pick#Documentation/git-cherry-pick.txt---emptydropkeepstop).

niik commented:
> Maybe we can just swap for `--empty=keep`

I think the problem is that we can't use either one in conjunction with `--continue`. I presume it's enough to keep it in `cherryPick` but not in `continueCherryPick`

niik committed:
Shuffle shuffle

mjcheetham commented on app/src/ui/preferences/advanced.tsx:
```diff
@@ -114,6 +127,37 @@ export class Advanced extends React.Component<
+            >
+              <p>
+                Use{' '}
+                <LinkButton uri="https://github.com/git-ecosystem/git-credential-manager">
```
We have a `gh.io` short link if you'd prefer to use that, incase this path ever changes in the future.

https://gh.io/gcm

sergiou87 replied:
Pinging @niik: in case you have missed this ☝️

niik replied:
:+1: Fixed in 2aa54a9e117ccc7dd86286127d44fbe911359173

mjcheetham commented on app/src/lib/git/credential.ts:
```diff
@@ -0,0 +1,52 @@
+
+const execCredential = (args: string[], path: string, stdin?: string) =>
+  // Can't use git() as that will call withTrampolineEnv which calls this method
+  GitProcess.exec(
+    [
+      ...['-c', 'credential.helper='],
+      ...['-c', `credential.helper=manager`],
+      'credential',
+      ...args,
+    ],
+    path,
+    {
+      stdin,
+      env: {
+        GIT_TERMINAL_PROMPT: '0',
+        GIT_ASKPASS: '',
+        TERM: 'dumb',
+      },
+    }
+  ).then(r => {
+    if (r.exitCode !== 0) {
+      // TODO: can we handle this error like we handle other errors in git()?
+      throw new Error(r.stderr)
+    }
+    return parseCredential(r.stdout)
+  })
```
Desktop ships with mingit on Windows, which includes GCM so `-c credential.helper=manager` should always be successful.

What happens on macOS and Linux when the user has not installed GCM? I'd imagine a generic error from Git as:

```shell
% git -c credential.helper= -c credential.helper=manager credential fill
git: 'credential-manager' is not a git command. See 'git --help'.
```

How does this surface to the user?

niik replied:
Great question. GitHub Desktop ships with its own Git environment (https://github.com/desktop/dugite) where we bundle GCM for macOS and Linux (https://github.com/desktop/dugite-native/pull/510).

niik committed:
Include username in credential url

In case there's more than one user per host and trampoline session

niik committed:
Merge branch 'development' into credential-helper

niik committed:
Extract findAccount and urlWithoutCredentials for sharing with credential helper

niik committed:
Split findAccount into two

niik committed:
Get rid of credential helper stuff in askpass handler (now legacy)

niik committed:
Add support for parsing stdin

niik committed:
Support credential helper identifier

niik committed:
Copy new trampolines in build

niik committed:
String or url, it's all the same

niik committed:
Bump trampoline

niik committed:
Bump trampoline again

niik committed:
Add flag for internal credential helper mode

niik committed:
Add helper function analogous with setGenericUsername + setGenericPassword

niik committed:
Only use setMostRecentGenericGitCredential in askpass handler

niik committed:
Change import

niik committed:
Initial credential helper mode

niik committed:
Support acquiring credential when not using external helper

niik committed:
:art: cleanup

niik committed:
:art: cleanup

niik committed:
Don't prompt for creds in background ops

niik committed:
Don't request credentials from external helper for github.com

niik committed:
Keep this on development for a bit longer

niik committed:
:book:

niik commented:
@sergiou87 @tidy-dev This is ready for a look. As we talked about this now adds a credential helper mode for our trampoline. There's a few rough edges mostly around cancelling auth where we no longer present the wall of text dialog that explains how to sign in again and instead show the raw error message from Git which is usually something like `fatal: Could not read Username for...`. As such I've kept this feature flagged to development so that we can focus on fixing the remaining bits in separate PRs (this has grown far too big already and you have my apologies for that).

@mjcheetham If you have some spare cycles to build this branch and test it out that'd be really appreciated as well!

niik committed:
Move OAuth dance into sign in store, create promise that encompasses entire sign in flow

niik committed:
Make sign in store methods non blocking

sergiou87 commented on app/src/ui/preferences/advanced.tsx:
```diff
@@ -114,6 +127,37 @@ export class Advanced extends React.Component<
+                  Git Credential Manager{' '}
+                </LinkButton>{' '}
+                for private repositories outside of GitHub.com. This feature is
+                experimental and subject to change{' '}
```
Nit: this can be removed, right?
```suggestion
                experimental and subject to change
```

sergiou87 commented on app/src/lib/trampoline/trampoline-environment.ts:
```diff
@@ -147,16 +210,27 @@ function deleteMostRecentGenericCredential(token: string) {
+  )
+}
+
+/** Returns the path of the desktop-askpass-trampoline binary. */
```
```suggestion
/** Returns the path of the desktop-credential-helper-trampoline binary. */
```

niik committed:
Remove unused state map

niik committed:
Throw authentication failure instead of prompts-disabled error

sergiou87 reviewed (commented):
Looks great! Take a look at the few comments, other than that I'd say it's ready to ship!

niik committed:
Don't allow GCM to be interactive when in background task

niik committed:
Detect GitHub hosts and prompt for sign in

niik committed:
Don't attempt to fetch metadata for GHE

niik committed:
Use shortlink to gcm web presence

niik committed:
The vastness of space

niik committed:
Update app/src/lib/trampoline/trampoline-environment.ts

Co-authored-by: Sergio Padrino <sergio.padrino@gmail.com>

niik committed:
Support arrays in credentials in the most hacky way imaginable

One day we'll replace Map with a proper Credential type which deals with this properly but for now let's just deal with it

niik committed:
This is neat

niik committed:
Endpoint discovery

niik committed:
Avoid managing GitHub creds in GCM

niik committed:
Play it safe and disable interactivity on all operations

niik committed:
:fire: Unused method

sergiou87 reviewed (dismissed):
Beautiful work!!

niik committed:
Request the meta endpoint

We exclude ghe.com so we can request it for GHES and even if it goes away in the future the header will still be there and we'll avoid a redirect

niik committed:
Might as well single this out, we only care about enterprise v non-enterprise at the moment

niik committed:
More exclusions

niik committed:
Don't hardcode GCM stuff into credential.ts

niik committed:
Don't depend on the auth type, realm is enough

niik committed:
:book:

niik committed:
:art: cleanup

niik committed:
:art: cleanup and :book:

niik committed:
Brain fart

niik committed:
Fix and test numbering of array entries

niik committed:
Defer loading accounts

Co-Authored-By: Sergio Padrino <1083228+sergiou87@users.noreply.github.com>

niik committed:
Merge pull request #18773 from desktop/prompt-for-sign-in

Prompt to sign in to GitHub hosts in credential helper

niik committed:
Merge branch 'development' into credential-helper

jaisansorachai reviewed (commented):
จะเอายังไง

Kavyasree221 reviewed (commented):
It is good
